src/backend/crew_ai/cleaned_llm_wrapper.py

```
@@ -45,6 +108,56 @@ def __init__(self, *args, **kwargs):
+            )
+            max_retries = 3
+        
+        for attempt in range(max_retries + 1):
+            try:
+                return super().call(messages, *args, **kwargs)
+            except Exception as e:
+                if not DynamicRateLimitHandler.is_rate_limit_error(e):
+                    raise  # Re-raise non-rate-limit errors
+                
+                if attempt >= max_retries:
+                    logger.error(f"❌ Rate limit: Max retries ({max_retries}) exceeded")
```

Ruff (0.15.6)

[warning] Use `logging.exception` instead of `logging.error`

Replace with `exception`

(TRY400)